packages/firestore/src/core/sync_engine_impl.ts

```diff
@@ -445,6 +475,27 @@ export async function syncEngineUnlisten(
   }
 }
 
+export async function triggerRemoteStoreUnlisten(
+  syncEngine: SyncEngine,
+  query: Query
+): Promise<void> {
+  const syncEngineImpl = debugCast(syncEngine, SyncEngineImpl);
+  const queryView = syncEngineImpl.queryViewsByQuery.get(query)!;
+  debugAssert(
+    !!queryView,
+    'Trying to unlisten on query not found:' + stringifyQuery(query)
+  );
+  const queries = syncEngineImpl.queriesByTarget.get(queryView.targetId)!;
+
+  // PORTING NOTE: Unregister the target ID with local Firestore client as
+  // watch target.
+  syncEngineImpl.sharedClientState.removeLocalQueryTarget(queryView.targetId);
+
+  if (queries.length === 1) {
```

Comment:
Should you check for primary client here too?

Also, i think `queries.length === 1` is always `true` because this method is only called when it is the last unlisten?

Reply:
`const queries = syncEngineImpl.queriesByTarget.get(queryView.targetId)!;` the queries could be mirror queries that has same target and targetId.  
we could have query1 listen to server, query2 listen to both cache and server. when query 2 unlisten to server, we should not close watch connection as query1 is still using it.